pkg/cmd/roachtest/main.go

```diff
@@ -296,6 +296,44 @@ func testsToRun(
 	return selectSpecs(notSkipped, selectProbability, true, print), nil
 }
 
+// updateSpecForSelectiveTests is responsible for updating the test spec skip and skip details
+// based on the test categorization criteria.
+func updateSpecForSelectiveTests(ctx context.Context, specs []registry.TestSpec) {
+	selectedTestsCount := 0
+	// run and select 35% of successful tests which gives a window of 3 days for all tests to run
+	selectedTests, err := testselector.CategoriseTests(ctx,
+		testselector.NewDefaultSelectTestsReq(35, roachtestflags.Cloud, roachtestflags.Suite))
+	if err != nil {
+		fmt.Printf("running all tests! error selecting tests: %v\n", err)
+		return
+	}
+	tdMap := make(map[string]*testselector.TestDetails)
+	for _, td := range selectedTests {
+		tdMap[td.Name] = td
+	}
+	for i := range specs {
+		if testShouldBeSkipped(tdMap, specs[i], roachtestflags.Suite) {
+			specs[i].Skip = "test selector"
+			specs[i].SkipDetails = "test skipped because it is stable and selective-tests is set."
+		} else {
+			selectedTestsCount++
+		}
+	}
+	fmt.Printf("%d out of %d tests selected for the run!\n", selectedTestsCount, len(specs))
+}
+
+// testShouldBeSkipped decides whether a test should be skipped based on test details and suite
+func testShouldBeSkipped(
+	testNamesToRun map[string]*testselector.TestDetails, test registry.TestSpec, suite string,
+) bool {
+	for test.TestSelectionOptOutSuites.IsInitialized() && test.TestSelectionOptOutSuites.Contains(suite) {
+		// test should not be skipped for this suite
+		return false
+	}
+	td, ok := testNamesToRun[test.Name]
+	return ok && test.Skip == "" && !td.Selected
```

Comment:
It seems strange to me that test selection would take into account tests that have been manually marked as skipped (`test.Skip != ""`). My understanding is that these skipped tests would then be selected to run (since they haven't run in a few days, presumably), but then immediately be skipped again since they are manually skipped.

Shouldn't test selection select tests among those that have a chance of running? Skipped tests don't have a chance of running from the get go.

Reply:
The query does not have a way to identify the skipped tests as all test statues are marked as "UNKNOWN". So, it selects all the tests that have been skipped and at this level, actually skipped tests are kept unchanged.

Reply:
I see, makes sense.

We can solve this in future work, but I still think it's not ideal that every manually skipped test (i.e., tests that someone marked with a `Skip` explanation on the `TestSpec`) will be selected (as they won't have run in a while), and then they will all be trivially skipped since they can't run.

Maybe the query itself can take a list of tests to include (tests where `Skip == ""`) or a list of tests to exclude (`Skip != ""`). But as I said can be done in future work.